src/coreclr/vm/comdelegate.cpp

```diff
@@ -789,8 +825,69 @@ LoaderHeap *DelegateEEClass::GetStubHeap()
     return GetInvokeMethod()->GetLoaderAllocator()->GetStubHeap();
 }
 
+#if defined(TARGET_RISCV64)
+static Stub* CreateILDelegateShuffleThunk(MethodDesc* pDelegateMD, bool callTargetWithThis)
+{
+    SigTypeContext typeContext(pDelegateMD);
+    MetaSig sig(pDelegateMD);
+    if (LoggingOn(LF_STUBS, LL_INFO1000000))
+    {
+        SString delegateName;
+        pDelegateMD->GetFullMethodInfo(delegateName);
+        LOGALWAYS(("CreateILDelegateShuffleThunk %s (%i args, callTargetWithThis:%i)\n",
+            delegateName.GetUTF8(), sig.NumFixedArgs(), callTargetWithThis));
+    }
+    _ASSERTE(sig.HasThis());
+
+    Module* pModule = sig.GetModule();
+    Signature signature = pDelegateMD->GetSignature();
+
+    ILStubLinkerFlags flags = ILSTUB_LINKER_FLAG_STUB_HAS_THIS;
+    ILStubLinker stubLinker(pModule, signature, &typeContext, pDelegateMD, flags);
+    ILCodeStream *pCode = stubLinker.NewCodeStream(ILStubLinker::kDispatch);
 
-static Stub* SetupShuffleThunk(MethodTable* pDelMT, MethodDesc* pTargetMeth)
+    for (unsigned i = 0; i < sig.NumFixedArgs(); ++i)
+        pCode->EmitLDARG(i);
+
+    pCode->EmitLoadThis();
+    pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__DELEGATE__METHOD_PTR_AUX)));
+    pCode->EmitCALLI(TOKEN_ILSTUB_TARGET_SIG, sig.NumFixedArgs(), sig.IsReturnTypeVoid() ? 0 : 1);
+    pCode->EmitRET();
+
+    MethodDesc* pStubMD = ILStubCache::CreateAndLinkNewILStubMethodDesc(
+        pDelegateMD->GetLoaderAllocator(), pDelegateMD->GetMethodTable(), ILSTUB_DELEGATE_SHUFFLE_THUNK,
+        pModule, signature.GetRawSig(), signature.GetRawSigLen(), &typeContext, &stubLinker);
+
+    // Build target signature
+    SigBuilder sigBuilder(signature.GetRawSigLen());
+    sigBuilder.AppendByte(callTargetWithThis
+        ? IMAGE_CEE_CS_CALLCONV_DEFAULT_HASTHIS
+        : IMAGE_CEE_CS_CALLCONV_DEFAULT);
+
+    unsigned numFixedArgs = sig.NumFixedArgs() - callTargetWithThis;
+    sigBuilder.AppendData(numFixedArgs);
+
+    SigPointer pReturn = sig.GetReturnProps();
+    pReturn.ConvertToInternalExactlyOne(pModule, &typeContext, &sigBuilder);
+
+    sig.SkipArg(); // skip delegate object
+    if (callTargetWithThis)
+        sig.SkipArg();
+
+    SigPointer pArgs = sig.GetArgProps();
+    for (unsigned i = 0; i < numFixedArgs; ++i)
+        pArgs.ConvertToInternalExactlyOne(pModule, &typeContext, &sigBuilder);
+
+    DWORD cbTargetSig;
+    PCCOR_SIGNATURE pTargetSig = (PCCOR_SIGNATURE)sigBuilder.GetSignature(&cbTargetSig);
+    ILStubResolver* pResolver = pStubMD->AsDynamicMethodDesc()->GetILStubResolver();
+    pResolver->SetStubTargetMethodSig(pTargetSig, cbTargetSig);
+
+    return Stub::NewStub(JitILStub(pStubMD), NEWSTUB_FL_SHUFFLE_THUNK);
```

Comment:
```suggestion
    return Stub::NewStub(JitILStub(pStubMD));
```
The `NEWSTUB_FL_...` flags are only meaningful for non-IL based stubs. It is probably harmless to set it for IL-based stubs, but it won't do anything useful either.

Reply:
`NEWSTUB_FL_SHUFFLE_THUNK` flag is used to provide good debugging experience. It ends up activating this codepath: https://github.com/dotnet/runtime/blob/667c0073fb8adb7ada0d55e89dc198d77f4b2a12/src/coreclr/vm/stubmgr.cpp#L1279-L1284

To provide the equivalent for IL-based shuffle thunks, could you please add something like this:
```
    else if (pStubMD->IsShuffleThunk())
    {
        LOG((LF_CORDB, LL_INFO1000, "ILSM::TraceManager: Delegate Shuffle Thunk\n"));
        return TraceShuffleThunk(trace, pContext, pRetAddr);
    }
```
to https://github.com/dotnet/runtime/blob/667c0073fb8adb7ada0d55e89dc198d77f4b2a12/src/coreclr/vm/stubmgr.cpp#L1909

Reply:
done